### ISSUE TITLE
POC: Handle dynamic cluster configuration

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#12851]
### Problem
Currently Zeebe brokers gets the cluster configuration from environment variables or application properties defined in yaml. This is static values. Once the cluster is started, we cannot change it. If a broker restarts with a different cluster configuration, this can lead to issues as well. 

Cluster configuration includes cluster related properties such clusterSize, partition count, replication factor, partition distribution etc. The configurations like segmentSize, network port etc. are not relevant as they could be broker specific and it is safe to restart broker with new values.

### Goal
To support dynamic scaling of Zeebe cluster, we first have to support dynamic configuration. 
- Initial cluster configuration will be provided via application properties or env variables as it is now. Once the cluster is started, the configuration must be persisted somewhere.
- If a broker restarts, it should use the persisted configuration instead of the static values configured in application configuration properties.
- The changes to cluster configuration must be made available to all brokers in a correct way. For example, when a new broker is added all brokers should learn about this with out much delay.

### Scope
- Identify which configuration properties are required to be made dynamic and persisted in the cluster. To limit the scope, consider this in the context of dynamic scaling.
- Develop a technical concept on how to manage cluster configuration dynamically.
    - Ideas: using gossip to propagate configuration changes. or use a central system partition to manage configuration.
- Implement a prototype
    - we can add a new broker and thus change the clusterSize. But not move the partitions to this broker. After restarting the whole cluster, it should be in the same state as before - with new clusterSize, new broker is part of the cluster, but partitions remain in the old brokers.

related to #4391 , https://github.com/camunda/product-hub/issues/1087

**Comment by lzgabel** [comment]
Looking forward to it!  🤩

**Comment by deepthidevaki** [comment]
Here I'm dumping some thoughts.

The cluster configuration that we are interested in can be defined as follows:

```
ClusterConfig = [BrokerConfig]
BrokerConfig = {
	nodeId: x
	state: { JOINING, ACTIVE, LEAVING }
	partitions: [
		1: state {JOINING, ACTIVE, LEAVING}
		2: ..
	]
}
```
We don't really have to keep clusterSize, partitionCount or replicationFactor explicitly. But we can infer it from the above configuration. 

Example for a cluster with one node. For simplicity, assume there is a central configuration.
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: ACTIVE
    }
  }
]
```
When a new broker is added, the new broker updates the configuration as follows:
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: ACTIVE
      }
  },
  {
    nodeId : 1
    state : JOINING
    partitions: {}
  }
]
```

After changing state to joining, and it is persisted, node 1 should start necessary operations to start join the cluster (what ever it is).

Once the node is fully joined the cluster, it marks its own state as active. 
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: ACTIVE
      }
  },
  {
    nodeId : 1
    state : ACTIVE
    partitions: {}
  }
]
```

Moving partition works as follows:
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: ACTIVE
      }
  },
  {
    nodeId : 1
    state : ACTIVE
    partitions: {
      2: JOINING
    }
  }
] 
```
After changing the state to joining and it is persisted, node 1 partition 2 should initiate necessary operations to join the partition.
If the node restarts while the state is Joining, it should re-do the operation necessary to join the partition.

Once it is fully joined the partition, the state is changed to Active.
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: ACTIVE
      }
  },
  {
    nodeId : 1
    state : ACTIVE
    partitions: {
      2: ACTIVE
    }
  }
] 
```
Then node 0 can leave partition 2, by first initiating leave operation.
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
        2: LEAVING
      }
  },
  {
    nodeId : 1
    state : ACTIVE
    partitions: {
      2: ACTIVE
    }
  }
]
```
Once it is fully left the partition:
```
[
  {
    nodeId : 0
    state: ACTIVE
    partitions: {
        1: ACTIVE
      }
  },
  {
    nodeId : 1
    state : ACTIVE
    partitions: {
      2: ACTIVE
    }
  }
]
```
On restart, node 0 reads it configuration, see that it has only one partition. And should start only raft thread for that partition.

if node 0 restarts while the state of partition 2 is leaving. It should do re-do any operations that it is supposed to do while leaving - for example start partition 2, and the send raft request to leave from the group., and get the confirmation that it left the partition gracefully.

**Comment by deepthidevaki** [comment]
Next question is where to persist the configuration. There are several options:

External :
1. Some central store (like how kafka used [Zookeeper](https://developer.confluent.io/learn-kafka/architecture/control-plane/#zookeeper-mode))
2. [Spring cloud config ](https://docs.spring.io/spring-cloud-config/docs/current/reference/html/)
3. Is it possible to use any k8s features for configuration management?

Internal:
Zeebe manages configuration itself. There are several options here:
1. A special raft partition for managing cluster configuration. We had a system partition before, when we first adopted atomix. [Kafka](https://developer.confluent.io/learn-kafka/architecture/control-plane/#kraft-mode-nascent) uses a special topic replicated to a subset of brokers for cluster management. 
2. Each broker persists the configuration locally, and propagates it via gossip. Systems like riak, cassandra follows a similar approach.

**Comment by Zelldon** [comment]
> The configurations like segmentSize, network port etc. are not relevant as they could be broker specific and it is safe to restart broker with new values.

I guess it makes sense related to segmentSize, since on replication we just write into our segment until it is full and create a new one. But what about the maxMessageSize, can this be an issue when it has different settings?

**Comment by deepthidevaki** [comment]
After considering different solutions, we decided to use gossip to store configuration in the cluster and handle configuration changes. 
Summary of our discussions:
- We prefer to keep the configuration management with in Zeebe to avoid the need for users to deploy and manage another system.
- We considered the use of a system partition which is replicated to a subset of brokers via raft protocol to manage configuration. While there are advantages to it, we decided against it due to how it can affect availability of the cluster.
- We also considered a using system partition together with gossip to offset the disadvantages of using only system partition. While this has the best of both worlds, it adds more complexity to the implementation.
- Gossip based solution works for our current use case, and is relatively easy to implement over current swim implementation. 

More details of the discussion is here [internal document](https://docs.google.com/document/d/1QiK6SKspSLTXy_S4yMIzbB4e2IyFGGrbx5Fmusrv8NQ/edit#heading=h.4kkpsdx96msa)

# Current proposed solution using gossip

**ClusterState** refers to the current configuration of the cluster. This includes information about which members are part of the cluster, and how partitions are distributed.

```
record ClusterState(long version, Map<MemberId, MemberState> members, ClusterChangePlan changes);
record MemberState(long version, State state, List<PartitionState> partitions);
record ClusterChangePlan(long version, List<Operation> operation);
record PartitionState(State state, int priority)
State = {JOINING, ACTIVE, LEAVING, LEFT}
```

**Member** - A node in the cluster. For configuration management, we are only interested in brokers and not gateways. So in this document, a member is always a broker.

**Coordinator** - A designated member which is the source of truth for the configuration. Only a coordinator can initiate configuration changes like add or remove members, or redistributing partitions. For the initial version, we assume node 0 is always the coordinator.

**Local configuration state** - Each member stores the known configuration in a local storage. 

**Gossip** - ClusterState is gossiped among cluster members using the existing SWIM implementation.

### Initialization Process
Each member is configured with an initial set of clusters. This is the same as the current configuration where the initial set is determined based on the clusterSize. We don’t have to change it.
- On bootstrap: local configuration state is empty
   - **Coordinator** :  First query initial set of cluster members for the current configuration. If no member replies with a known configuration, coordinator generates a new configuration from the provided static properties with version=0. Coordinator only has to get a response from any one of the members.
      - Optionally, we can disable this auto initialization in the coordinator. We will expose an api to trigger initialization which has to be explicitly invoked. This should be done only once when the cluster is created. In helm, it could be done by a k8s job which is executed once after the chart is installed.
   - **Member**: Other members request the coordinator for the configuration. Raft partitions and other zeebe services that depend on the cluster configuration are only started after this node receives the configuration from the coordinator.
- On restart
   - **Member/Coordinator**: All members  use the local persisted configuration.
   - If there are any pending operations in clusterState.changes, then apply it. (see below Making configuration changes).
- On restart with data loss
   - **Member/Coordinator**: Works similar to bootstrap.


### Making configuration changes
**Coordinator**: Only the coordinator can make configuration changes.  The coordinator first generates a ConfigChangePlan operation that contains the steps needed to achieve the final configuration.
Example:
```
[
 Partition 1: [ Node 4 Join, Node 3 Leave] // Move partition 1 from 3 to 4.
 Partition 2: [ Node 5 Join, Node 2 Leave, Node 6 Join, Node 1 Leave] // Move partition 2 from nodes 1 and 2 to nodes 5 and 6.
]
```
Adding new partitions can also be done similarly.

This config change plan is updated in ClusterState and gossiped as follows.

- New ClusterState is generated as follows:
   - Increment current clusterState.version
   - Add operations to ClusterState.changes
   - Set the ClusterState.changes.verion = new clusterState.version
- Updates the local configuration state
- Gossip new ClusterState

Coordinator also ensure that only one change is in progress.

**Member**: When a member receives a new ClusterState via gossip
- Updates its local configuration state by merge(localState, receivedState)
  - Merge operation resolves concurrent updates using versions.
- If changes.firstOperation().memberId == localMemberId
  - Apply operation
    - Execute necessary changes for the operation. This can be async and can take a long time to complete. For example when joining a partition it has to first start a raft partition, then it may have to request the leader to join the replication group and then wait until the raft log is caught up.
   - When the operation is completed generate new ClusterState as follows
     - Add new MemberState for the member to ClusterState if the operation is to join a new member
     - If changing existing members (such as adding a new partition), increment the version in MemberState and update the state. A member should only update its own state. And each update is marked by a version increment. This is to avoid and resolve conflicts.
     - Remove operation from ConfigChangePlan.
     - Update local configuration state
     - Gossip the new state. (The next operation will be applied when the corresponding member receives this update).
- If the member restarts during any of these steps, it is ok. The member can re-apply the operation. The operations should be idempotent. 

### How does this solution handle conflicts?
Only coordinator can initiate configuration changes. Thus only coordinator can update ClusterState.version. Each member uses this version to merge local cluster state and received cluster state.
Only a member can update its own state. Thus only a member can update MemberState.version. When merging, choose the state with the highest version.

### How does this solution handle restarts with data loss?
All members ask the coordinator for the configuration if it cannot find a local configuration. 
Coordinator always asks other known members, before generating its own configuration.

### How does this solution detect conflicting configurations?
All members use coordinator’s configuration. So there is no conflict. 
The only rare case is if the coordinator restarts with data loss and end up generating a different configuration. However chances of this are less since the coordinator always asks other known members before generating a configuration. If in rare case it happens, coordinator can detect inconsistent configuration via version number.

### How does this solution handle node leaving?
This is similar to other configuration changes. In graceful leaving, the member itself initiates the configuration change and sets its state as “LEFT”. Once the member is left, its info is still in the cluster state with state “LEFT”. We can run delayed garbage collection to permanently remove the node from the cluster state after a pre-defined duration. For non-graceful cases, the coordinator can update the member state as left to force remove the member.

**Comment by deepthidevaki** [comment]
I'm closing this issue. I will create a new issue to keep track of the implementation tasks.
